### ISSUE TITLE
ionic 4 No provider for Camera!

### ISSUE THREAD

**Issue by sreerajkarippala** [ionic-team/ionic#17917]
# Bug Report

**Ionic version:**

[x] **4.x**

**Current behavior:**

When a page call native plugin camera then it produce an error
```
ERROR Error: Uncaught (in promise): Error: StaticInjectorError(AppModule)[CreatePage -> Camera]: 
  StaticInjectorError(Platform: core)[CreatePage -> Camera]: 
    NullInjectorError: No provider for Camera!
Error: StaticInjectorError(AppModule)[CreatePage -> Camera]: 
  StaticInjectorError(Platform: core)[CreatePage -> Camera]: 
    NullInjectorError: No provider for Camera!
    at NullInjector.push../node_modules/@angular/core/fesm5/core.js.NullInjector.get (core.js:8895)
    at resolveToken (core.js:9140)
    at tryResolveToken (core.js:9084)
    at StaticInjector.push../node_modules/@angular/core/fesm5/core.js.StaticInjector.get (core.js:8981)
    at resolveToken (core.js:9140)
    at tryResolveToken (core.js:9084)
    at StaticInjector.push../node_modules/@angular/core/fesm5/core.js.StaticInjector.get (core.js:8981)
    at resolveNgModuleDep (core.js:21217)
    at NgModuleRef_.push../node_modules/@angular/core/fesm5/core.js.NgModuleRef_.get (core.js:21906)
    at resolveNgModuleDep (core.js:21217)
    at NullInjector.push../node_modules/@angular/core/fesm5/core.js.NullInjector.get (core.js:8895)
    at resolveToken (core.js:9140)
    at tryResolveToken (core.js:9084)
    at StaticInjector.push../node_modules/@angular/core/fesm5/core.js.StaticInjector.get (core.js:8981)
    at resolveToken (core.js:9140)
    at tryResolveToken (core.js:9084)
    at StaticInjector.push../node_modules/@angular/core/fesm5/core.js.StaticInjector.get (core.js:8981)
    at resolveNgModuleDep (core.js:21217)
    at NgModuleRef_.push../node_modules/@angular/core/fesm5/core.js.NgModuleRef_.get (core.js:21906)
    at resolveNgModuleDep (core.js:21217)
    at resolvePromise (zone.js:831)
    at resolvePromise (zone.js:788)
    at zone.js:892
    at ZoneDelegate.push../node_modules/zone.js/dist/zone.js.ZoneDelegate.invokeTask (zone.js:423)
    at Object.onInvokeTask (core.js:17289)
    at ZoneDelegate.push../node_modules/zone.js/dist/zone.js.ZoneDelegate.invokeTask (zone.js:422)
    at Zone.push../node_modules/zone.js/dist/zone.js.Zone.runTask (zone.js:195)
    at drainMicroTaskQueue (zone.js:601)
```

**Other information:**

I already insert provider in `app.module.ts`

```
import { Camera } from '@ionic-native/camera/ngx';
...
@NgModule({
  declarations: [AppComponent],
  entryComponents: [],
  imports: [
    BrowserModule,
    FormsModule,
    ReactiveFormsModule,
    IonicModule.forRoot(),
    AppRoutingModule,
    HttpClientModule,
    IonicStorageModule.forRoot(),
    LoanAvailablePageModule
  ],
  providers: [
    AppVersion,
    Device,
    StatusBar,
    SplashScreen,
    { provide: RouteReuseStrategy, useClass: IonicRouteStrategy },
    Firebase,
    HeaderColor,
    SocialSharing,
    ImagePicker,
    Camera
  ],
  bootstrap: [AppComponent]
})
```

**Ionic info:**

```
Ionic:

   ionic (Ionic CLI)             : 4.12.0 (/usr/local/lib/node_modules/ionic)
   Ionic Framework               : @ionic/angular 4.1.2
   @angular-devkit/build-angular : 0.13.6
   @angular-devkit/schematics    : 7.2.4
   @angular/cli                  : 7.3.6
   @ionic/angular-toolkit        : 1.4.1

Cordova:

   cordova (Cordova CLI) : 8.1.2 (cordova-lib@8.1.1)
   Cordova Platforms     : android 7.1.4
   Cordova Plugins       : cordova-plugin-ionic-keyboard 2.1.3, cordova-plugin-ionic-webview 3.1.2, (and 10 other plugins)

System:

   ios-deploy : 1.8.2
   ios-sim    : 5.0.2
   NodeJS     : v10.10.0 (/usr/local/bin/node)
   npm        : 6.4.1
   OS         : macOS Mojave
```

**Comment by liamdebeasi** [comment]
Hi there,

Thanks for the issue. Can you post the contents of your `package.json` as well as the TypeScript file for the page that calls the Camera plugin?

Thanks!

**Comment by sreerajkarippala** [comment]
sure, my `package.json`
```
{
  "name": "xapp",
  "version": "0.0.1",
  "author": "Ionic Framework",
  "homepage": "https://ionicframework.com/",
  "scripts": {
    "ng": "ng",
    "start": "ng serve",
    "build": "ng build",
    "test": "ng test",
    "lint": "ng lint",
    "e2e": "ng e2e"
  },
  "private": true,
  "dependencies": {
    "@angular/common": "^7.2.2",
    "@angular/core": "^7.2.2",
    "@angular/forms": "^7.2.2",
    "@angular/http": "^7.2.2",
    "@angular/platform-browser": "^7.2.2",
    "@angular/platform-browser-dynamic": "^7.2.2",
    "@angular/router": "^7.2.2",
    "@ionic-native/app-version": "^5.3.0",
    "@ionic-native/camera": "^5.3.0",
    "@ionic-native/core": "^5.0.0",
    "@ionic-native/device": "^5.3.0",
    "@ionic-native/firebase": "^5.3.0",
    "@ionic-native/header-color": "^5.3.0",
    "@ionic-native/image-picker": "^5.3.0",
    "@ionic-native/social-sharing": "^5.3.0",
    "@ionic-native/splash-screen": "^5.0.0",
    "@ionic-native/status-bar": "^5.3.0",
    "@ionic/angular": "^4.1.0",
    "@ionic/storage": "^2.2.0",
    "cordova-android": "7.1.4",
    "cordova-plugin-app-version": "0.1.9",
    "cordova-plugin-camera": "4.0.3",
    "cordova-plugin-device": "2.0.2",
    "cordova-plugin-firebase": "2.0.5",
    "cordova-plugin-headercolor": "1.0.0",
    "cordova-plugin-ionic-keyboard": "^2.1.3",
    "cordova-plugin-ionic-webview": "^3.1.2",
    "cordova-plugin-splashscreen": "^5.0.2",
    "cordova-plugin-statusbar": "2.4.2",
    "cordova-plugin-telerik-imagepicker": "2.3.0",
    "cordova-plugin-whitelist": "^1.3.3",
    "cordova-plugin-x-socialsharing": "5.4.4",
    "cordova-sqlite-storage": "3.2.0",
    "core-js": "^2.5.4",
    "es6-promise-plugin": "^4.2.2",
    "rxjs": "~6.3.3",
    "zone.js": "~0.8.29"
  },
  "devDependencies": {
    "@angular-devkit/architect": "~0.12.3",
    "@angular-devkit/build-angular": "~0.13.0",
    "@angular-devkit/core": "~7.2.3",
    "@angular-devkit/schematics": "~7.2.3",
    "@angular/cli": "~7.3.1",
    "@angular/compiler": "~7.2.2",
    "@angular/compiler-cli": "~7.2.2",
    "@angular/language-service": "~7.2.2",
    "@ionic/angular-toolkit": "~1.4.0",
    "@types/node": "~10.14.2",
    "@types/jasmine": "~2.8.8",
    "@types/jasminewd2": "~2.0.3",
    "codelyzer": "~4.5.0",
    "jasmine-core": "~2.99.1",
    "jasmine-spec-reporter": "~4.2.1",
    "karma": "~3.1.4",
    "karma-chrome-launcher": "~2.2.0",
    "karma-coverage-istanbul-reporter": "~2.0.1",
    "karma-jasmine": "~1.1.2",
    "karma-jasmine-html-reporter": "^0.2.2",
    "protractor": "~5.4.0",
    "ts-node": "~8.0.0",
    "tslint": "~5.12.0",
    "typescript": "~3.1.6"
  },
  "description": "An Ionic project",
  "cordova": {
    "plugins": {
      "cordova-plugin-app-version": {},
      "cordova-plugin-firebase": {},
      "cordova-plugin-device": {},
      "cordova-plugin-headercolor": {},
      "cordova-plugin-x-socialsharing": {},
      "cordova-plugin-statusbar": {},
      "cordova-sqlite-storage": {},
      "cordova-plugin-whitelist": {},
      "cordova-plugin-splashscreen": {},
      "cordova-plugin-ionic-webview": {
        "ANDROID_SUPPORT_ANNOTATIONS_VERSION": "27.+"
      },
      "cordova-plugin-ionic-keyboard": {},
      "cordova-plugin-telerik-imagepicker": {},
      "cordova-plugin-camera": {}
    },
    "platforms": [
      "android"
    ]
  }
}
```

**Comment by liamdebeasi** [comment]
Thanks! Can you also post the `.ts` file for the page your `ion-slides` are in?

**Comment by sreerajkarippala** [comment]
ion-slides? hope you looking for `.ts` source where camera plugin in used

```
import { Component, OnInit } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { ActionSheetController, Platform, ToastController } from '@ionic/angular';
import { ActivatedRoute, Router } from '@angular/router';

import { Camera, CameraOptions, PictureSourceType } from '@ionic-native/Camera/ngx';
import { ImagePicker, ImagePickerOptions } from '@ionic-native/image-picker/ngx';

import { FormGroup, FormControl, Validators, FormBuilder } from '@angular/forms';

import { SystemService } from '../../system/system.service';
import { environment } from '../../../environments/environment';
import { Observable } from 'rxjs';

@Component({
  selector: 'app-create',
  templateUrl: './create.page.html',
  styleUrls: ['./create.page.scss']
})
export class CreatePage implements OnInit {
  ...

  adForm: FormGroup;
  contactForm: FormGroup;
  additionalForm: FormGroup;

  constructor(
    private http: HttpClient,
    private system: SystemService,
    private platform: Platform,
    // private camera: Camera,
    private imagePicker: ImagePicker,
    private actionSheetController: ActionSheetController,
    private formBuilder: FormBuilder,
    public toastController: ToastController,
    private route: ActivatedRoute,
    private router: Router,
  ) { }

  ngOnInit() {
    this.getCategories();
    this.getLocations();
    const d = new Date();
    let currentYear = d.getFullYear();
    for (let index = 0; index < 100; index++) {
      this.made_yr.push(currentYear--);
    }

    this.adForm = this.formBuilder.group({
      category: ['', [Validators.required]],
      title: ['', [Validators.required, Validators.minLength(6)]],
      price: [],
      brand: ['', [Validators.required]],
      model: ['', [Validators.required]],
      made_year: ['', [Validators.required]],
      driven: [],
      fuel_type: [],
      desc: []
    });
    this.contactForm = this.formBuilder.group({
      location: ['', [Validators.required]],
      city: [],
      place: ['', [Validators.required]],
      email: [],
      phone: [
        '',
        [
          Validators.required,
          Validators.minLength(10),
          Validators.pattern(/^[0-9]/)
        ]
      ]
    });
    this.additionalForm = this.formBuilder.group({});
  }

  ....

  async actAddImage() {
    const actionSheet = await this.actionSheetController.create({
      header: 'Select Image source',
      buttons: [
        {
          text: 'Load from Library',
          handler: () => {
            // this.takePicture(this.camera.PictureSourceType.PHOTOLIBRARY);
            this.browsePictures();
          }
        },
         {
           text: 'Use Camera',
           handler: () => {
             this.takePicture(this.camera.PictureSourceType.CAMERA);
           }
         },
        {
          text: 'Cancel',
          role: 'cancel'
        }
      ]
    });
    await actionSheet.present();
  }

  browsePictures() {
    console.log('browsePictures(): clicked');
    this.uploadProgress = true;
    const options: ImagePickerOptions = {
      maximumImagesCount: 8,
      outputType: 1
    };
    this.imagePicker.getPictures(options).then((pickedImages) => {

      if (pickedImages.length > 0) {

        for (let index = 0; index < pickedImages.length; index++) {
          const imagesLength = this.images.length;

          this.images = this.images.concat({
            url: null,
            title: 'uploading...',
            is_loading: false
          });

          const headers = new HttpHeaders({
            'Content-Type': 'application/x-www-form-urlencoded;charset=UTF-8'
          });

          this.http.post(this.system.generateRequestUrl('/enterprise/ocm', null),
            this.system.encodeUri({
              image: 'data:image/png;base64,' + pickedImages[index],
              req: 'upload_image',
              app: 'ocm',
              key: this.system.API_KEY
            }),
            { headers: headers }
          ).subscribe((response: any) => {
            if (response.status === true) {
              console.log('uploadImage: this.images.length', this.images.length);
              this.images[imagesLength] = {
                url: this.ASSET_URL + response.data,
                title: 'Uploaded',
                is_loading: false
              };
              this.ad.ad_image = this.ad.ad_image.concat(response.data);
            }
            this.uploadProgress = false;

          }, (error: any) => {
            this.images[imagesLength] = {
              url: null,
              title: 'failed...',
              is_loading: false
            };
            this.uploadProgress = false;

          });
        }

      } else {
        this.uploadProgress = false;
      }
    });
  }

   takePicture(sourceType: PictureSourceType) {
     const options: CameraOptions = {
       quality: 100,
       sourceType: sourceType,
       saveToPhotoAlbum: true,
       correctOrientation: true,
       encodingType: this.camera.EncodingType.PNG,
       destinationType: this.camera.DestinationType.DATA_URL,
       mediaType: this.camera.MediaType.PICTURE
     };

     this.camera.getPicture(options).then(image => {
       console.log(image);

       if (
         this.platform.is('android') &&
         sourceType === this.camera.PictureSourceType.PHOTOLIBRARY
       ) {
         this.uploadImage(image);
       } else {
         this.uploadImage(image);
       }
    });
   }
...
}
```

**Comment by liamdebeasi** [comment]
Ha, sorry! I did want the `.ts` file, but I had another issue in my mind (where I was also looking at the `.ts` file).

Thanks! I'll take a look and get back to you soon.

**Comment by liamdebeasi** [comment]
Hi there,

I took a look at your code. Your code is importing the camera module incorrectly.

Your code has:

```
import { Camera, CameraOptions, PictureSourceType } from '@ionic-native/Camera/ngx';
``` 

but it should be:

```
import { Camera, CameraOptions, PictureSourceType } from '@ionic-native/camera/ngx';
```

Note the lowercase `c` in `@ionic-native/camera/ngx`. I tried the incorrect import and got the error you reported originally. I am going to close this issue as using the correct import fixes it.

Thanks for using Ionic!
